lib/router/src/cache.rs

```diff
@@ -1,426 +1,321 @@
-                    capacity: self.capacity,
-                });
-            }
-        }
-
-        Ok(Reserve {
-            vals: &mut self.vals,
-            now: &self.now,
-        })
-    }
-
-    /// Overrides the time source for tests.
-    #[cfg(test)]
-    fn with_clock<M: Now>(self, now: M) -> Cache<K, V, M> {
-        Cache {
-            now,
-            vals: self.vals,
-            capacity: self.capacity,
-            max_idle_age: self.max_idle_age,
-        }
-    }
-}
-
-// ===== impl Reserve =====
-
-impl<'a, K: Hash + Eq + 'a, V: 'a, N: Now + 'a> Reserve<'a, K, V, N> {
-    /// Stores a route in the cache.
-    pub fn store(self, key: K, val: V) {
-        let node = Node::new(val.into(), self.now.now());
-        self.vals.insert(key, node);
-    }
-}
-
-// ===== impl Access =====
+            expires,
+            expirations: DelayQueue::with_capacity(capacity),
+            values: IndexMap::default(),
+        };
+        let cache = Lock::new(cache);
+        let bg_purge = PurgeCache {
+            cache: cache.clone(),
+            interval: Interval::new_interval(expires),
+        };
 
-impl<'a, T: 'a, N: Now + 'a> Deref for Access<'a, T, N> {
-    type Target = T;
-    fn deref(&self) -> &Self::Target {
-        &self.node
+        (cache, bg_purge)
     }
-}
 
-impl<'a, T: 'a, N: Now + 'a> DerefMut for Access<'a, T, N> {
-    fn deref_mut(&mut self) -> &mut Self::Target {
-        &mut self.node
+    pub fn capacity(&self) -> usize {
+        self.capacity
     }
-}
 
-impl<'a, T: 'a, N: Now + 'a> Access<'a, T, N> {
-    #[cfg(test)]
-    fn last_access(&self) -> Instant {
-        self.node.last_access
+    pub fn can_insert(&self) -> bool {
+        self.values.len() < self.capacity
     }
-}
 
-impl<'a, T: 'a, N: Now + 'a> Drop for Access<'a, T, N> {
-    fn drop(&mut self) {
-        self.node.last_access = self.now.now();
-    }
-}
-
-// ===== impl Node =====
+    /// Attempts to access an item by key.
+    ///
+    /// If a value is returned, this key will not be considered for eviction
+    /// for another `expires` span of time.
+    pub fn access(&mut self, key: &K) -> Option<V> {
+        if let Some(node) = self.values.get_mut(key) {
+            self.expirations.reset(&node.dq_key, self.expires);
+            trace!("reset expiration for cache value associated with key");
+
+            return Some(node.value.clone());
+        }
 
-impl<T> Node<T> {
-    pub fn new(value: T, last_access: Instant) -> Self {
-        Node { value, last_access }
+        None
     }
 
-    pub fn access<'a, N: Now + 'a>(&'a mut self, now: &'a N) -> Access<'a, T, N> {
-        Access { now, node: self }
-    }
+    /// Attempts to insert an item by key.
+    ///
+    /// If a value is returned, this key has been set to expire after an
+    /// `expires` span of time.
+    pub fn insert(&mut self, key: K, value: V) -> Option<V> {
+        let node = {
+            trace!("inserting value into expirations");
```

Comment:
Similarly, I'd suggest that we move this up to debug and make the phrasing uniform with the expiration:

```rust
debug!("inserting an item from the cache")
```